refl1d/lib_numba/magnetic.py

```diff
@@ -0,0 +1,397 @@
+    "E0": numba.float64,
+    "L": numba.int32,
+    "LP": numba.int32,
+    "STEP": numba.int8,
+    "Z": numba.float64,
+}
+CR4XA_LOCALS.update((s, numba.complex128) for s in [
+    "S1L", "S1LP", "S3", "S3LP", 
+    "FS1S1", "FS1S3", "FS3S1", "FS3S3",
+    "DELTA", "BL", "GL", "BLP", "GLP", "SSWAP", "DETW",
+    "DBB", "DBG", "DGB", "DGG",
+    "ES1L", "ENS1L", "ES1LP", "ENS1LP", "ES3L", "ENS3L", "ES3LP", "ENS3LP"])
+CR4XA_LOCALS.update(("A{i}{j}".format(i=i, j=j), numba.complex128) for i in range(1,5) for j in range(1,5))
+CR4XA_LOCALS.update(("B{i}{j}".format(i=i, j=j), numba.complex128) for i in range(1,5) for j in range(1,5))
+CR4XA_LOCALS.update(("C{i}".format(i=i), numba.complex128) for i in range(1,5))
+
+@numba.njit(CR4XA_SIG, parallel=False, cache=True, locals=CR4XA_LOCALS)
+def Cr4xa(N, D, SIGMA, IP, RHO, IRHO, RHOM, U1, U3, KZ, Y):
+    EPS = 1e-10
+
+    if (KZ <= -1.e-10):
+        L = N-1
+        STEP = -1
+        SIGMA_OFFSET = -1
+    elif (KZ >= 1.e-10):
+        L = 0
+        STEP = 1
+        SIGMA_OFFSET = 0
+    else:
+        Y[0] = -1.
+        Y[1] = 0.
+        Y[2] = 0.
+        Y[3] = -1.
+        return
+
+    #    Changing the target KZ is equivalent to subtracting the fronting
+    #    medium SLD.
+
+    # IP = 1 specifies polarization of the incident beam I+
+    # IP = -1 specifies polarization of the incident beam I-
+    E0 = KZ*KZ + PI4*(RHO[L]+IP*RHOM[L])
+
+    Z = 0.0
+    if (N > 1):
+        # chi in layer 1
+        LP = L + STEP
+        # Branch selection:  the -sqrt below for S1 and S3 will be
+        #     +Imag for KZ > Kcrit,
+        #     -Real for KZ < Kcrit
+        # which covers the S1, S3 waves allowed by the boundary conditions in the
+        # fronting and backing medium:
+        # either traveling forward (+Imag) or decaying exponentially forward (-Real).
+        # The decaying exponential only occurs for the transmitted forward wave in the backing:
+        # the root +iKz is automatically chosen for the incident wave in the fronting.
+        #
+        # In the fronting, the -S1 and -S3 waves are either traveling waves backward (-Imag)
+        # or decaying along the -z reflection direction (-Real) * (-z) = (+Real*z).
+        # NB: This decaying reflection only occurs when the reflected wave is below Kcrit
+        # while the incident wave is above Kcrit, so it only happens for spin-flip from
+        # minus to plus (lower to higher potential energy) and the observed R-+ will
+        # actually be zero at large distances from the interface.
+        #
+        # In the backing, the -S1 and -S3 waves are explicitly set to be zero amplitude
+        # by the boundary conditions (neutrons only incident in the fronting medium - no
+        # source of neutrons below).
+        #
+        S1L = -sqrt(complex(PI4*(RHO[L]+RHOM[L]) - E0, -PI4*(fabs(IRHO[L])+EPS)))
+        S3L = -sqrt(complex(PI4*(RHO[L]-RHOM[L]) -
+                    E0, -PI4*(fabs(IRHO[L])+EPS)))
```

Comment:
Break after comma to keep expressions together.

Why does this code use `-PI4...` while the code in `lib/magnetic.cc` was changed in this PR to use `+PI4...`?

Reply:
This is an excellent question, and I wish I had written down what I was doing.

Reply:
I think I was testing the sign conventions, and never reverted the .cc file because it was no longer being used. The negative signs in this expression are designed to explicitly choose the branch for the sqrt. There might be a better way to do this.